### PR TITLE
Add De/Serialization for CompactionInput/Result

### PR DESCRIPTION
**Pull request by jay-zhuang**
Summary: The functions will be used for remote compaction parameter
input and result.

Test Plan: `make check`

### CODE REVIEW

File: db/compaction/compaction_job_test.cc [hunk 1]
```diff
@@ -1096,6 +1096,150 @@ TEST_F(CompactionJobTest, OldestBlobFileNumber) {
...
+
+  std::string output;
+  Status s = input.Write(&output);
+  ASSERT_OK(s);
+
+  // Test deserialization
+  CompactionServiceInput deserialized1;
+  s = CompactionServiceInput::Read(output, &deserialized1);
+  ASSERT_OK(s);
+  ASSERT_TRUE(deserialized1.TEST_Equals(&input));
```
[the remark this answers is not in the document]
Comment (jay-zhuang) [reply db/compaction/compaction_job_test.cc:1135]: Currently comparator and other pointer options are not deserialized. So for the first version, the user has to set them on the remote side.
The pointer options deserialization will be implemented soon, so we could remove the requirement of user setting these.

[the remark this answers is not in the document]
Comment (ajkr) [reply db/compaction/compaction_job_test.cc:1141]: I see. It'd be nice if we provided a way to register RocksDB's builtin extensions - but again, outside the scope of this PR.

File: db/compaction/compaction_job.cc [hunk 2]
```diff
@@ -2062,4 +2067,433 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
...
+     {0, OptionType::kBinaryString, OptionVerificationType::kNormal,
+      OptionTypeFlags::kNone}},
+    {"options",
+     {offset_of(&ColumnFamilyDescriptor::options), OptionType::kConfigurable,
+      OptionVerificationType::kNormal, OptionTypeFlags::kNone,
+      [](const ConfigOptions& opts, const std::string& /*name*/,
+         const std::string& value, char* addr) {
+        auto conf = CFOptionsAsConfigurable(ColumnFamilyOptions());
+        auto status = conf->ConfigureFromString(opts, value);
+        auto cf_options = reinterpret_cast<ColumnFamilyOptions*>(addr);
```
[the remark this answers is not in the document]
Comment (ajkr) [reply db/compaction/compaction_job.cc:2108]: Got it thanks. Sorry I didn't mean that you need to change it - it's outside the scope of this PR.

[the remark this answers is not in the document]
Comment (mrambacher) [reply db/compaction/compaction_job.cc:2108]: @ajkr @jay-zhuang I do not know if void* or char* is more appropriate.  When I implemented these functions originally, I kept the same types (char*) as was used in the original code (example from 6.5):
extern bool ParseOptionHelper(char* opt_address, const OptionType& opt_type,
                                          const std::string& value);
bool SerializeSingleOptionHelper(const char* opt_address,
                                 const OptionType opt_type, std::string* value);

If you believe it should be a void*, I am fine with changing it (and would rather it be sooner than later before there are too many more place to touch).  

And in the scope of a different PR.

File: db/compaction/compaction_job.cc [hunk 3]
```diff
@@ -2062,4 +2068,439 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
...
+            this_conf->AreEquivalent(opts, that_conf.get(), &mismatch_opt);
+        if (!result) {
+          *mismatch = name + "." + mismatch_opt;
+        }
+        return result;
+      }}},
+};
+
+static std::unordered_map<std::string, OptionTypeInfo> cs_input_type_info = {
+    {"column_family", OptionTypeInfo::Struct("column_family", &cfd_type_info, 0,
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:2142]: The 0 should probably be offset_of, right?  If someone changes the struct, won't this break?

File: db/compaction/compaction_job.cc [hunk 4]
```diff
@@ -2062,4 +2068,439 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
...
+
+static CompactionServiceInput dummy_cs_input;
+template <typename T1>
+int offset_of(T1 CompactionServiceInput::*member) {
+  return int(size_t(&(dummy_cs_input.*member)) - size_t(&dummy_cs_input));
+}
+
+static std::unordered_map<std::string, OptionTypeInfo> cfd_type_info = {
+    {"name",
+     {0, OptionType::kBinaryString, OptionVerificationType::kNormal,
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:2099]: 0 should be offset of

File: db/compaction/compaction_job.cc [hunk 5]
```diff
@@ -2062,4 +2068,439 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
...
+      OptionTypeFlags::kNone}},
+    {"options",
+     {offset_of(&ColumnFamilyDescriptor::options), OptionType::kConfigurable,
+      OptionVerificationType::kNormal, OptionTypeFlags::kNone,
+      [](const ConfigOptions& opts, const std::string& /*name*/,
+         const std::string& value, char* addr) {
+        auto conf = CFOptionsAsConfigurable(ColumnFamilyOptions());
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:2106]: This code might be simpler if you use GetColumnFamilyOptionsFromString (in convenience.h)

File: db/compaction/compaction_job.cc [hunk 6]
```diff
@@ -2062,4 +2068,439 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
...
+            OptionsHelper::kCFOptionsName);
+        return status;
+      },
+      [](const ConfigOptions& opts, const std::string& /*name*/,
+         const char* addr, std::string* value) {
+        const auto cf_options =
+            reinterpret_cast<const ColumnFamilyOptions*>(addr);
+        auto conf = CFOptionsAsConfigurable(*cf_options);
+        std::string result;
+        auto status = conf->GetOptionString(opts, &result);
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:2119]: GetStringFromColumnFamilyOptions?

File: db/compaction/compaction_job.cc [hunk 7]
```diff
@@ -2062,4 +2068,439 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
...
+                                             OptionTypeFlags::kNone)},
+    {"db_options",
+     {offset_of(&CompactionServiceInput::db_options), OptionType::kConfigurable,
+      OptionVerificationType::kNormal, OptionTypeFlags::kNone,
+      [](const ConfigOptions& opts, const std::string& /*name*/,
+         const std::string& value, char* addr) {
+        auto options = reinterpret_cast<DBOptions*>(addr);
+        auto conf = DBOptionsAsConfigurable(*options);
+        auto status = conf->ConfigureFromString(opts, value);
+        *options = *conf->GetOptions<DBOptions>(OptionsHelper::kDBOptionsName);
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:2153]: GetDBOptionsFromString

File: db/compaction/compaction_job.cc [hunk 8]
```diff
@@ -2062,4 +2068,439 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
...
+        return status;
+      },
+      [](const ConfigOptions& opts, const std::string& /*name*/,
+         const char* addr, std::string* value) {
+        const auto options = reinterpret_cast<const DBOptions*>(addr);
+        auto conf = DBOptionsAsConfigurable(*options);
+        std::string result;
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:2160]: GetStringFromDBOptions

File: db/compaction/compaction_job.cc [hunk 9]
```diff
@@ -2062,4 +2068,439 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
...
+        {"marked_for_compaction",
+         {offsetof(struct CompactionServiceOutputFile, marked_for_compaction),
+          OptionType::kBoolean, OptionVerificationType::kNormal,
+          OptionTypeFlags::kNone}},
+};
+
+static std::unordered_map<std::string, OptionTypeInfo>
+    compaction_job_stats_type_info = {
+        {"elapsed_micros",
+         {0, OptionType::kUInt64T, OptionVerificationType::kNormal,
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:2249]: offsetof

File: db/compaction/compaction_job.cc [hunk 10]
```diff
@@ -2062,4 +2068,439 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
...
+          OptionTypeFlags::kNone}},
+        {"num_single_del_mismatch",
+         {offsetof(struct CompactionJobStats, num_single_del_mismatch),
+          OptionType::kUInt64T, OptionVerificationType::kNormal,
+          OptionTypeFlags::kNone}},
+};
+
+static std::unordered_map<std::string, OptionTypeInfo> cs_result_type_info = {
+    {"output_files",
+     OptionTypeInfo::Vector<CompactionServiceOutputFile>(
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:2366]: offsetof

File: db/compaction/compaction_job.cc [hunk 11]
```diff
@@ -2062,4 +2068,439 @@ void CompactionServiceCompactionJob::CleanupCompaction() {
...
+     {offsetof(struct CompactionServiceResult, bytes_written),
+      OptionType::kUInt64T, OptionVerificationType::kNormal,
+      OptionTypeFlags::kNone}},
+    {"stats", OptionTypeInfo::Struct(
+                  "stats", &compaction_job_stats_type_info,
+                  offsetof(struct CompactionServiceResult, stats),
+                  OptionVerificationType::kNormal, OptionTypeFlags::kNone)},
+};
+
+namespace {
```
Comment (mrambacher) [inline db/compaction/compaction_job.cc:2400]: With PR8249, you (hopefully) should not need these classes any longer.

File: db/compaction/compaction_job.h [hunk 12]
```diff
@@ -285,6 +299,15 @@ struct CompactionServiceResult {
   uint64_t bytes_read;
   uint64_t bytes_written;
   CompactionJobStats stats;
+
+  // serialization interface to read and write the object
+  static Status Read(const std::string& data_str, CompactionServiceResult* obj);
+  Status Write(std::string* output);
```
Comment (mrambacher) [inline db/compaction/compaction_job.h:305]: Should Write take a BinaryFormatVersion of what format to write?

File: db/compaction/compaction_job_test.cc [hunk 13]
```diff
@@ -1096,6 +1096,156 @@ TEST_F(CompactionJobTest, OldestBlobFileNumber) {
...
+  output.erase(pos, to_remove.length());
+  s = CompactionServiceResult::Read(output, &deserialized3);
+  ASSERT_OK(s);
+  mismatch.clear();
+  ASSERT_FALSE(deserialized3.TEST_Equals(&result, &mismatch));
+  ASSERT_EQ(mismatch, "bytes_read");
+
+  deserialized3.bytes_read = 123;
+  ASSERT_TRUE(deserialized3.TEST_Equals(&result));
+}
```
Comment (mrambacher) [inline db/compaction/compaction_job_test.cc:1247]: Can you add a test where the version in the string is not valid?

**Review by mrambacher — COMMENTED** [review]
This generally looks very good.

One general question is if there is a need for any sort of checksum on the serialized strings?

File: options/options_helper.cc [hunk 14]
```diff
@@ -539,6 +539,11 @@ static bool ParseOptionHelper(char* opt_address, const OptionType& opt_type,
       return ParseEnum<CompactionStopStyle>(
           compaction_stop_style_string_map, value,
           reinterpret_cast<CompactionStopStyle*>(opt_address));
+    case OptionType::kBinaryString: {
```
[the remark this answers is not in the document]
Comment (jay-zhuang) [reply options/options_helper.cc:481]: sounds good, changed to `kEncodedString`.

File: db/compaction/compaction_job_test.cc [hunk 15]
```diff
@@ -1096,6 +1096,148 @@ TEST_F(CompactionJobTest, OldestBlobFileNumber) {
...
+  ASSERT_TRUE(deserialized2.TEST_Equals(&input));
+
+  // Test missing field
+  CompactionServiceInput deserialized3;
+  deserialized3.output_level = 0;
+  std::string to_remove = "output_level=4;";
+  size_t pos = output.find(to_remove);
+  ASSERT_TRUE(pos != std::string::npos);
+  output.erase(pos, to_remove.length());
+  s = CompactionServiceInput::Read(output, &deserialized3);
```
[the remark this answers is not in the document]
Comment (jay-zhuang) [reply db/compaction/compaction_job_test.cc:1167]: It's `assert_ok` the next line. maybe good to combine these 2 lines. changed.

File: db/compaction/compaction_job.cc (hunk 3 above)
Comment (jay-zhuang) [reply db/compaction/compaction_job.cc:2142]: sure, I'm just following the other existing structure definition.

File: db/compaction/compaction_job.h (hunk 12 above)
Comment (jay-zhuang) [reply db/compaction/compaction_job.h:305]: We can add that if we have other format versions.

File: db/compaction/compaction_job.cc [hunk 16]
```diff
@@ -1984,9 +1987,11 @@ Status CompactionServiceCompactionJob::Run() {
       c->column_family_data()->CalculateSSTWriteHint(c->output_level());
   bottommost_level_ = c->bottommost_level();
 
-  compact_->sub_compact_states.emplace_back(c, compaction_input_.begin,
-                                            compaction_input_.end,
-                                            compaction_input_.approx_size);
+  Slice begin = compaction_input_.begin;
+  Slice end = compaction_input_.end;
+  compact_->sub_compact_states.emplace_back(
+      c, begin.size() > 0 ? &begin : nullptr, end.size() > 0 ? &end : nullptr,
```
[the remark this answers is not in the document]
Comment (jay-zhuang) [reply db/compaction/compaction_job.cc]: Added info if `begin` and `end` is nullptr or not.

**Pull request closed without merging**